src/System.Private.CoreLib/shared/System/Threading/WaitHandle.cs

```diff
@@ -0,0 +1,384 @@
+            bool success = false;
+            try
+            {
+                for (int i = 0; i < waitHandles.Length; ++i)
+                {
+                    WaitHandle waitHandle = waitHandles[i];
+                    if (waitHandle == null)
+                    {
+                        throw new ArgumentNullException("waitHandles[" + i + ']', SR.ArgumentNull_ArrayElement);
+                    }
+
+                    SafeWaitHandle safeWaitHandle = waitHandle._waitHandle;
+                    if (safeWaitHandle == null)
+                    {
+                        // Throw ObjectDisposedException for backward compatibility even though it is not be representative of the issue
+                        throw new ObjectDisposedException(null, SR.ObjectDisposed_Generic);
+                    }
+
+                    safeWaitHandle.DangerousAddRef(ref success);
+                    safeWaitHandles[i] = safeWaitHandle;
+                    unsafeWaitHandles[i] = safeWaitHandle.DangerousGetHandle();
+                }
+                success = true;
+            }
+            finally
+            {
+                if (!success)
+                {
+                    for (int i = 0; i < waitHandles.Length; ++i)
+                    {
+                        SafeWaitHandle safeWaitHandle = safeWaitHandles[i];
+                        if (safeWaitHandle == null)
+                        {
+                            break;
+                        }
+                        safeWaitHandle.DangerousRelease();
+                        safeWaitHandles[i] = null;
+                    }
+                }
+            }
+        }
+
+        private static int WaitMultiple(ReadOnlySpan<WaitHandle> waitHandles, bool waitAll, int millisecondsTimeout)
+        {
+            if (waitHandles == null)
+            {
+                throw new ArgumentNullException(nameof(waitHandles), SR.ArgumentNull_Waithandles);
+            }
+            if (waitHandles.Length == 0)
+            {
+                //
+                // Some history: in CLR 1.0 and 1.1, we threw ArgumentException in this case, which was correct.
+                // Somehow, in 2.0, this became ArgumentNullException.  This was not fixed until Silverlight 2,
+                // which went back to ArgumentException.
+                //
+                // Now we're in a bit of a bind.  Backward-compatibility requires us to keep throwing ArgumentException
+                // in CoreCLR, and ArgumentNullException in the desktop CLR.  This is ugly, but so is breaking
+                // user code.
+                //
+                throw new ArgumentException(SR.Argument_EmptyWaithandleArray, nameof(waitHandles));
+            }
+            if (waitHandles.Length > MaxWaitHandles)
+            {
+                throw new NotSupportedException(SR.NotSupported_MaxWaitHandles);
+            }
+            if (-1 > millisecondsTimeout)
```

Comment:
nit: swap the operands order

Reply:
Yeah, I started fixing it up, but didn't do it at all places yet. Looks like someone took their C habits to C# (putting to constant on the left side prevents accidentally writing `=` instead of `==`) ;-)